src/mono/mono/eventpipe/ep-rt-mono-runtime-provider.c

```diff
@@ -0,0 +1,4586 @@
+
+			gc_heap_dump_context_buffer_free (context->buffer);
+
+			if (context->bulk_type_logger)
+				bulk_type_event_logger_free (context->bulk_type_logger);
+		}
+
+		memset (context, 0, sizeof (GCHeapDumpContext));
+	}
+}
+
+static
+GCHeapDumpContext *
+gc_heap_dump_context_alloc (
+	EVENTPIPE_TRACE_CONTEXT trace_context,
+	uint32_t gc_reason,
+	uint32_t gc_type,
+	uint32_t gc_count,
+	uint32_t gc_depth)
+{
+	GCHeapDumpContext *context = g_new0 (GCHeapDumpContext, 1);
+	gc_heap_dump_context_init (
+		context,
+		trace_context,
+		gc_reason,
+		gc_type,
+		gc_count,
+		gc_depth);
+
+	return context;
+}
+
+static
+void
+gc_heap_dump_context_free (GCHeapDumpContext *context)
+{
+	gc_heap_dump_context_fini (context);
+	g_free (context);
+}
+
+static
+GCHeapDumpContext *
+gc_heap_dump_context_get (void)
+{
+	EventPipeMonoThreadData *thread_data = ep_rt_mono_thread_data_get_or_create ();
+	return thread_data ? thread_data->gc_heap_dump_context : NULL;
+}
+
+static
+void
+gc_heap_dump_context_set (GCHeapDumpContext *context)
+{
+	EventPipeMonoThreadData *thread_data = ep_rt_mono_thread_data_get_or_create ();
+	if (thread_data) {
+		if (thread_data->gc_heap_dump_context)
+			gc_heap_dump_context_free (thread_data->gc_heap_dump_context);
+		thread_data->gc_heap_dump_context = context;
+	}
+}
+
+static
+void
+gc_heap_dump_context_reset (void)
+{
+	gc_heap_dump_context_set (NULL);
+}
+
+static
+int32_t
+gc_roots_sort_compare_func (
+	const void *a,
+	const void *b)
+{
+	EP_ASSERT (a && b);
+
+	GCRootData *root_a = *(GCRootData **)a;
+	GCRootData *root_b = *(GCRootData **)b;
+
+	EP_ASSERT (root_a && root_b);
+	return root_a->start - root_b->start;
+}
+
+static
+void
+gc_heap_dump_context_build_roots (GCHeapDumpContext *context)
+{
+	EP_ASSERT (is_gc_heap_dump_enabled (context));
+
+	ep_rt_spin_lock_requires_lock_not_held (&_gc_lock);
+
+	EP_SPIN_LOCK_ENTER (&_gc_lock, section1)
+		if (!context->gc_roots) {
+			dn_vector_ptr_custom_alloc_params_t params = { 0, };
+			params.capacity = dn_umap_size (&_gc_roots_table);
+			context->gc_roots = dn_vector_ptr_custom_alloc (&params);
+
+		} else {
+			dn_vector_ptr_clear (context->gc_roots);
+		}
+
+		DN_UMAP_FOREACH_BEGIN (const mono_byte *, key, GCRootData *, value, &_gc_roots_table) {
+			DN_UNREFERENCED_PARAMETER (key);
+			dn_vector_ptr_push_back (context->gc_roots, value);
+		} DN_UMAP_FOREACH_END;
+	EP_SPIN_LOCK_EXIT (&_gc_lock, section1)
+
+	dn_vector_ptr_sort (context->gc_roots, gc_roots_sort_compare_func);
+
+ep_on_exit:
+	ep_rt_spin_lock_requires_lock_not_held (&_gc_lock);
+	return;
+
+ep_on_error:
+	ep_exit_error_handler ();
+}
+
+static
+GCRootData *
+gc_heap_dump_context_find_root (
+	GCHeapDumpContext *context,
+	uintptr_t root)
+{
+	EP_ASSERT (is_gc_heap_dump_enabled (context));
+
+	const uint8_t *base = (const uint8_t *)dn_vector_ptr_data (context->gc_roots);
+	const uint8_t *p;
+	size_t lim;
+	int32_t cmp;
+	for (lim = dn_vector_ptr_size (context->gc_roots); lim; lim >>= 1) {
```

Comment:
Lets add it to the list of things to do. I inlined it here mainly to get away from doing function calls to compare function since we will search the list for each root.